src/coreclr/vm/exceptionhandling.cpp

```diff
@@ -3121,18 +3063,19 @@ PropagateForeignExceptionThroughNativeFrames(IN     PEXCEPTION_RECORD   pExcepti
 
 #endif // HOST_WINDOWS
 
-extern "C" void * QCALLTYPE CallCatchFunclet(QCall::ObjectHandleOnStack exceptionObj, BYTE* pHandlerIP, REGDISPLAY* pvRegDisplay, ExInfo* exInfo)
+void CallCatchFunclet(OBJECTREF throwable, BYTE* pHandlerIP, REGDISPLAY* pvRegDisplay, ExInfo* exInfo)
 {
-    QCALL_CONTRACT;
-
-    BEGIN_QCALL;
-    GCX_COOP_NO_DTOR();
-
+    CONTRACTL
+    {
+        MODE_COOPERATIVE;
+        GC_NOTRIGGER;
```

Comment:
If it's not too much trouble, what's the reasoning behind this function being GC_NOTRIGGER and then having GC contract violations in it? Compatibility with existing code?

Reply:
The point was that the whole function should not trigger GC, but the actual catch handler is managed code, so obviously GC can happen in it. So the contract violation helped me to express that.
However, I am changing that in my upcoming commit after hitting some issue with profiler and thinking again about when the GC is prohibited. Looking at the legacy EH code, the GC is prohibited there only in-between finallys in the same frame. That makes sense and I am moving to that model. Without it, profiler notifications would be a problem, as each frame we visit sends a profiler notification and that switches to preemptive GC mode, thus GC can happen there. So we cannot prohibit GC for the whole 2nd pass.